Spring for GraphQL and React (#1444)
oktadeveloper/okta-blog: _source/_posts/2023-07-27-spring-graphql-react.md

---
layout: blog_post
title: "How to build a GraphQL API with Spring for GraphQL"
author: jimena-garbarino
by: contractor
communities: [security,java,javascript]
description: "A step-by-step guide for building a secured GraphQL API with Spring Boot and Auth0 authentication on React"
tags: [java, javascript]
tweets:
- ""
- ""
- ""
image:

> type: awareness

Reviewer: ```suggestion
type: awareness
github: https://github.com/oktadev/auth0-spring-graphql-react-example
```

This won't show up on the Auth0 blog, but it's useful when previewing it here.

This step-by-step guide demonstrates how to build a GraphQL API with Spring Boot and Spring for GraphQL, for querying a sample dataset of related companies, persons, and properties, seeded to a Neo4j database. It also demonstrates how to build a React client with Next.js and MUI Datagrid to consume the API. Both the client and the server are secured with Auth0 authentication using Okta Spring Boot Starter for the server and Auth0 React SDK for the client.

{% img blog/spring-graphql-react/logos.png alt:"Spring, GraphQL and React" width:"700" %}{: .center-image }

> This example was created with the following tools and services:
> - [Node.js v18.16.1](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm)
> - [npm 9.5.1](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm)
> - [Java OpenJDK 17](https://jdk.java.net/java-se-ri/17)
> - [Docker 24.0.2](https://docs.docker.com/desktop/)
> - [Auth0 account](https://auth0.com/signup)
> - [Auth0 CLI 1.0.0](https://github.com/auth0/auth0-cli#installation)
> - [HTTPie 3.2.2](https://httpie.io/)
> - [Next.js 13.4.19](https://nextjs.org/)

{% include toc.md %}

> ## Build a GraphQL API with Spring for GraphQL

Reviewer: I think we should add a section before this that shows how to get started quickly. That is, clone the repo, configure it for Auth0, and run it. If the repo's README has everything, you could condense it to a single paragraph. Or you could follow something similar to what I did in my Angular and Spring Boot post. https://auth0.com/blog/spring-boot-angular-crud/

That way, when this tutorial stops working in a year or two (because the Spring Boot version is no longer available on start.spring.io), people can still see the app running.

Author: I was thinking the same while writing the post, I like the structure of the angular post. For this time, I just added a note below the logos image, linking to the repository.

        this.countryOfOrigin = countryOfOrigin;
    }

    public LocalDate getIncorporationDate() {
        return incorporationDate;
    }

    public void setIncorporationDate(LocalDate incorporationDate) {
        this.incorporationDate = incorporationDate;
    }

    public Integer getMortgagesOutstanding() {
        return mortgagesOutstanding;
    }

    public void setMortgagesOutstanding(Integer mortgagesOutstanding) {
        this.mortgagesOutstanding = mortgagesOutstanding;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }
}
```

Create the package `com.okta.developer.demo.repository` and the class `CompanyRepository`:

__CompanyRepository.java__
```java
package com.okta.developer.demo.repository;

import com.okta.developer.demo.domain.Company;
import org.springframework.data.neo4j.repository.ReactiveNeo4jRepository;

public interface CompanyRepository extends ReactiveNeo4jRepository<Company, Long> {

}
```

> Create the configuration class `GraphQLConfig` under the root package. This class will enable CORS from the React client and log the GraphQL schema mappings:

Reviewer: Should this say "This class will (eventually) enable CORS" since it doesn't do it yet?

Author: I moved the cors config to properties and added the comment there.

__GraphQLConfig.java__
```java
package com.okta.developer.demo;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.autoconfigure.graphql.GraphQlSourceBuilderCustomizer;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration(proxyBeanMethods = false)
class GraphQLConfig {

    private static Logger logger = LoggerFactory.getLogger("graphql");

    @Bean
    public GraphQlSourceBuilderCustomizer sourceBuilderCustomizer() {
        return (builder) ->
                builder.inspectSchemaMappings(report -> {
                    logger.debug(report.toString());
                });
    }
}
```

> Create the configuration class `SpringBootGraphQlApiConfig` in the root package as well, defining a reactive transaction manager required for reactive Neo4j:

Reviewer: ```suggestion
Create a configuration class named `SpringBootGraphQlApiConfig` in the root package as well, defining a reactive transaction manager required for reactive Neo4j:
```

Reviewer: I think we should be consistent as to how you capitalize GraphQL. In the previous class, you have an uppercase L and here you have a lowercase L.

Author: I did it t match the application class created by spring initializr. Maybe I should change the application name too?

                .variable("page", 0)
                .execute()
                .path("companyList")
                .matchesJson("""
                    [{
                        "id": null,
                        "SIC": "1234",
                        "name": "Test Company",
                        "status": "active",
                        "category": "private",
                        "companyNumber": "12345678",
                        "countryOfOrigin": "UK"
                    }]
                """);
    }
}
```

Create the document file `companyList.graphql` containing the query definition for the test, in the folder `src/main/test/resources/graphql-test`:

__companyList.graphql__
```graphql
query companyList($page: Int) {
    companyList(page: $page) {
        id
        SIC
        name
        status
        category
        companyNumber
        countryOfOrigin
    }
}
```

Update the test configuration in `build.gradle` file, so passed tests are logged:

__build.gradle__
```groovy
tasks.named('test') {
    useJUnitPlatform()

    testLogging {
        // set options for log level LIFECYCLE
        events "failed", "passed"
    }
}
```

Run the test with:

```shell

> ./gradlew clean test

Reviewer: ```suggestion
./gradlew test
```

There's no need for `clean` here since you have compiled anything yet.

```

You should see logs for the successful tests:

```
...
SpringBootGraphQlApiApplicationTests > contextLoads() PASSED

CompanyControllerTests > shouldGetCompanies() PASSED
...
```

### Add Neo4j seed data

> Let's add Neo4j migrations dependency for the seed data insertion. Edit the file `build.gradle` and add:

Reviewer: ```suggestion
Let's add Neo4j migrations dependency for the seed data insertion. Edit the `build.gradle` file and add:
```

__build.gradle__
```groovy
dependencies {
    ...

>     implementation 'eu.michael-simons.neo4j:neo4j-migrations-spring-boot-starter:2.5.3'

Reviewer: ```suggestion
    implementation 'eu.michael-simons.neo4j:neo4j-migrations-spring-boot-starter:2.5.3'
    ...
```

Add an ellipsis after because there are more `testImplementation` dependencies, or remove the dependencies wrapper.

MATCH (c:Company {companyNumber: row.` CompanyNumber`})
SET c.name = row.CompanyName,
c.mortgagesOutstanding = toInteger(row.`Mortgages.NumMortOutstanding`),
c.incorporationDate = Date(Datetime({epochSeconds: apoc.date.parse(row.IncorporationDate,'s','dd/MM/yyyy')})),
c.SIC = row.`SICCode.SicText_1`,
c.countryOfOrigin = row.CountryOfOrigin,
c.status = row.CompanyStatus,
c.category = row.CompanyCategory;
```

__V006__Land.cypher__
```cypher
LOAD CSV WITH HEADERS FROM "file:///LandOwnershipAmericans.csv" AS row
MATCH (c:Company {companyNumber: row.`Company Registration No. (1)`})
MERGE (p:Property {titleNumber: row.`Title Number`})
SET p.address = row.`Property Address`,
p.county  = row.County,
p.price   = toInteger(row.`Price Paid`),
p.district = row.District
MERGE (c)-[r:OWNS]->(p)
WITH row, c,r,p WHERE row.`Date Proprietor Added` IS NOT NULL
SET r.date = Date(Datetime({epochSeconds: apoc.date.parse(row.`Date Proprietor Added`,'s','dd-MM-yyyy')}));
CREATE INDEX FOR (c:Company) ON c.incorporationDate;
```
{% endraw %}

Update `application.properties` and add the following properties:

__application.properties__
```properties

> ...

Reviewer: Since there's no data in this file yet, I don't think the `...` line is necessary.

Author: Removed

spring.graphql.graphiql.enabled=true
spring.graphql.schema.introspection.enabled=true
org.neo4j.migrations.transaction-mode=PER_STATEMENT
spring.neo4j.uri=bolt://localhost:7687
spring.neo4j.authentication.username=neo4j

spring.graphql.cors.allowed-origins=http://localhost:3000
```

Create a `.env` file in the project root to store the Neo4j credentials:

__.env__
```shell

> export SPRING_NEO4J_AUTHENTICATION_PASSWORD=verysecret

Reviewer: I wish Spring Boot supported .env files by default so you don't have to `source` this file first. I recently used [spring-dotenv](https://github.com/paulschwarz/spring-dotenv) for this. 

No need to change to use this, just an FYI.

Reviewer: If it was possible to use a dotenv equivalent, you could just have one file in the root directory for Spring Boot and Neo4j in Docker. Then, you could start the Neo4j images with `docker compose -f src/main/docker/neo4j.yml`.

Reviewer: In case you weren't aware, Spring Boot 3.1 has Docker Compose support built-in. This means you can add a dependency and a compose file in the root directory, and Spring Boot will auto-start it. https://spring.io/blog/2023/06/21/docker-compose-support-in-spring-boot-3-1

I used it in this example I created last week: https://github.com/oktadev/auth0-java-microservices-examples/tree/main/spring-boot-gateway-webflux/car-service

```

> Download the following seed files to some folder:

Reviewer: ```suggestion
Download the following seed files to some folder:
```
If they need to be moved from some folder to somewhere else, it might be a good idea to specify that now.

Author: I added a comment suggesting to use an empty folder.

- [CompanyDataAmericans](https://guides.neo4j.com/ukcompanies/data/CompanyDataAmericans.csv)
- [LandOwnershipAmericans](https://guides.neo4j.com/ukcompanies/data/LandOwnershipAmericans.csv)

> - [PSCAmericans.csv](https://guides.neo4j.com/ukcompanies/data/PSCAmericans.csv)

Reviewer: The last file has a `.csv` suffix in its link. We should be consistent and remove it from the last one or add it to the first two.

Author: Added de missing suffix

Create the folder `src/main/docker` and create a file `neo4j.yml` there, with the following content:

__neo4j.yml__
```yml

> # This configuration is intended for development purpose, it's **your** responsibility to harden it for production

Reviewer: This line makes me think you copied it from JHipster. Since this is just an example, I think it's OK to remove it.

Author: Removed.

name: companies
services:
  neo4j:
    image: neo4j:5
    volumes:
      - <csv-folder>:/var/lib/neo4j/import
    environment:
      - NEO4J_AUTH=neo4j/${NEO4J_PASSWORD}
      - NEO4JLABS_PLUGINS=["apoc"]
    # If you want to expose these ports outside your dev PC,
    # remove the "127.0.0.1:" prefix
    ports:
      - '127.0.0.1:7474:7474'
      - '127.0.0.1:7687:7687'
    healthcheck:
      test: ['CMD', 'wget', 'http://localhost:7474/', '-O', '-']
      interval: 5s
      timeout: 5s
      retries: 10
```

Create the file `src/main/docker/.env` with the following content:

__.env__

> ```shell

Reviewer: ```suggestion
```dotenv
```

NEO4J_PASSWORD=verysecret
```

As you can see the compose file will mount `<csv-folder>` to a `/var/lib/neo4j/import` volume, making the content accessible from the running neo4j container.

> Replace `<csv-folder>` with the path to the CSV files downloaded before.

Reviewer: For me, the `<csv-folder>` is `~/Downloads`, so that's what I put in `neo4j.yml`. I believe this is the default on macOS and Linux.

Author: I think it should be a clean folder, will add a comment.

Author: Added a comment to use an empty folder.

In a terminal, go to the `docker` folder and run:

```shell
docker compose -f neo4j.yml up
```

### Run the API server

Go to the project root folder and start the application with:

```shell
source .env && ./gradlew bootRun
```

> Wait for the logs to inform the seed data migrations have run (it might take a while):

Reviewer: Why does it take so long? You might elaborate on that.

Author: There was a missing constraint, originally only suported by enterprise version. I added the constraint and timing improved.

```
2023-08-02T13:06:14.386-03:00  INFO 28673 --- [           main] a.s.neo4j.migrations.core.Migrations     : Applied migration 001 ("Constraint").
2023-08-02T13:06:23.379-03:00  INFO 28673 --- [           main] a.s.neo4j.migrations.core.Migrations     : Applied migration 002 ("Company").
2023-08-02T13:11:23.693-03:00  INFO 28673 --- [           main] a.s.neo4j.migrations.core.Migrations     : Applied migration 003 ("Person").
2023-08-02T13:21:03.680-03:00  INFO 28673 --- [           main] a.s.neo4j.migrations.core.Migrations     : Applied migration 004 ("PersonCompany").
2023-08-02T13:21:06.519-03:00  INFO 28673 --- [           main] a.s.neo4j.migrations.core.Migrations     : Applied migration 005 ("CompanyData").
2023-08-02T13:21:06.551-03:00  INFO 28673 --- [           main] a.s.neo4j.migrations.core.Migrations     : Applied migration 006 ("Land").

> ```

Reviewer: I was going to suggest replacing the first part of each line with `...`, but then I realized the timestamps are a good indicator of how long it takes. Maybe do something like this so folks don't have to scroll right to see the good stuff:

```
2023-08-02T13:06:14.386-03:00  ... Applied migration 001 ("Constraint").
2023-08-02T13:06:23.379-03:00  ... Applied migration 002 ("Company").
2023-08-02T13:11:23.693-03:00  ... Applied migration 003 ("Person").
2023-08-02T13:21:03.680-03:00  ... Applied migration 004 ("PersonCompany").
2023-08-02T13:21:06.519-03:00  ... Applied migration 005 ("CompanyData").
2023-08-02T13:21:06.551-03:00  ... Applied migration 006 ("Land").
```

Author: The original queries had a missing constraint that was causing the slow inserts. I added a constraint and it improved. Also formatted the logs.

> Test the API with GraphiQL at **http://localhost:8080/graphiql**. In the query box on the left, paste the following query

Reviewer: ```suggestion
Test the API with GraphiQL at `http://localhost:8080/graphiql`. In the query box on the left, paste the following query:
```
Just following the convention we've used for localhost URLs over the years. I'm not sure that the Auth0 cares as much as the OktaDev blog. I'm just trying to be consistent.

```graphql
{
    companyList(page: 20) {
        id
        SIC
        name
        status
        category
        companyNumber
        countryOfOrigin
    }
}
```

You should see the query output in the box on the right:

{% img blog/spring-graphql-react/graphiql-test.png alt:"GraphiQL example" width:"900" %}{: .center-image }

> __NOTE:__

> > If you see a warning message in the server logs, that reads _The query used a deprecated function: id_, you can ignore it, Spring Data Neo4j still [behaves correctly](https://github.com/spring-projects/spring-data-neo4j/issues/2716).

Reviewer: I like that you linked to the issue! It looks like it might have a solution at the bottom that people can use, too. I tried adding the following to `SpringBootGraphQlApiConfig`, but I still see the error:

```java
    @Bean
    org.neo4j.cypherdsl.core.renderer.Configuration cypherDslConfiguration() {
        return org.neo4j.cypherdsl.core.renderer.Configuration.newConfig().withDialect(Dialect.NEO4J_5).build();
    }
```

## Build a React client

> Now let's create a Single Page Application (SPA) to consume the GraphQL API with React and Next.js. The list of companies will display in a [MUI](https://mui.com/material-ui/getting-started/) [Datagrid](https://mui.com/x/react-data-grid/) component. The application will use Next.js `app` router. The `src/app` folder will contain routing files only, and the UI components and application code will be located in other folders.

Reviewer: ```suggestion
Now let's create a Single Page Application (SPA) to consume the GraphQL API with React and Next.js. The list of companies will display in a [MUI](https://mui.com/material-ui/getting-started/) [Data Grid](https://mui.com/x/react-data-grid/) component. The application will use Next.js' `app` router. The `src/app` folder will only contain routing files, and the UI components and application code will be in other folders.
```

Install Node and in a terminal run:

```shell

> npx create-next-app@13.4.19

Reviewer: Should I run this in the root of the Spring Boot app or a directory alongside it? I could look at the GitHub repo to figure it out, but I think you should say it here to reduce ambiguity.

Author: Added a comment.

```

Answer the questions as follows:

```
✔ What is your project named? ... react-graphql
✔ Would you like to use TypeScript? ... Yes
✔ Would you like to use ESLint? ... Yes
✔ Would you like to use Tailwind CSS? ... No
✔ Would you like to use `src/` directory? ... Yes
✔ Would you like to use App Router? (recommended) ... Yes
✔ Would you like to customize the default import alias? ... No
```

Then add the the MUI Datagrid dependencies and custom hooks from Vercel:

```shell
cd react-graphql && \
  npm install @mui/x-data-grid && \
  npm install @mui/material@5.14.5 @emotion/react @emotion/styled && \
  npm install react-use-custom-hooks
```

> Test run the application with:

Reviewer: ```suggestion
Run the application with:
```

```shell
npm run dev
```

> Navigate to http://localhost:3000 and you should see the default Next.js page:

Reviewer: ```suggestion
Navigate to `http://localhost:3000` and you should see the default Next.js page:
```

{% img blog/spring-graphql-react/nextjs-default.png alt:"Next.js default page" width:"800" %}{: .center-image }

### Create the API client

Create the folder `src/services` and add the file `base.tsx` with the following code:

{% raw %}
__base.tsx__
```tsx

> import axios from "axios";

Reviewer: I typically use single quotes in TSX files, and double quotes for the HTML part, like JHipster does. Are you using double quotes for any particular reason?

Reviewer: Should Axios be installed at this point? I'm getting errors from IntelliJ, and I don't see it in `package.json`.

Author: Added the axios dependency. I was using double quotes because the quick fix of my IDE was configured by default with double quotes. Replaced by single quotes.

      return response.data.data.companyCount as number;
    } catch (error) {
      console.log("handle get company count error", error);
      if (error instanceof AxiosError) {
        let axiosError = error as AxiosError;
        if (axiosError.response?.data) {
          throw new Error(axiosError.response?.data as string);
        }
      }
      throw new Error("Unknown error, please contact the administrator");
    }
  },

  getCompanyList: async (params?: CompaniesQuery) => {
    try {
      const response = await backendAPI.post("/graphql", {
        query: `{
        companyList(page: ${params?.page || 0}) {
          name,
          SIC,
          id,
          companyNumber,
          category
        }}`,
      });
      return response.data.data.companyList as CompanyDTO[];
    } catch (error) {
      console.log("handle get companies error", error);
      if (error instanceof AxiosError) {
        let axiosError = error as AxiosError;
        if (axiosError.response?.data) {
          throw new Error(axiosError.response?.data as string);
        }
      }
      throw new Error("Unknown error, please contact the administrator");
    }
  },

};
```
{% endraw %}

> Add a file `.env.example` and `.env.local` in the root folder, both with the following content:

Reviewer: Why are both necessary? And should one be added to `.gitignore`?

Author: Added a note about the files.

    : 1;

  return (
    <>
      <Box>
        <Typography variant="h4" component="h1">
          Companies
        </Typography>
      </Box>
      <Box mt={2}>
        <CompanyTableContainer page={page}></CompanyTableContainer>
      </Box>
    </>
  );
};

export default HomePage;
```
{% endraw %}

Replace the contents of `src/app/page.tsx` and change it to render the `HomePage` component:

{% raw %}
__app/page.tsx__
```tsx
import HomePage from "./HomePage";

const Page = () => {
  return (
    <HomePage></HomePage>
  );
}

export default Page;
```
{% endraw %}

Add a component defining the page width, for using it in the root layout. Create `src/layout/WideLayout.tsx` with the following content:

{% raw %}
__WideLayout.tsx__
```tsx
"use client";

import { Container, ThemeProvider, createTheme } from "@mui/material";

const theme = createTheme({
  typography: {
    fontFamily: "inherit",
  },
});

const WideLayout = (props: { children: React.ReactNode }) => {
  return (
    <ThemeProvider theme={theme}>
      <Container maxWidth="lg" sx={{ mt: 4 }}>
        {props.children}
      </Container>
    </ThemeProvider>
  );
};

export default WideLayout;
```
{% endraw %}

> With the implementation above, the page content will be wrapped in a `ThemeProvider` component, so https://github.com/vercel/next.js/discussions/45433[MUI child components inherit the font family] from the root layout.

Reviewer: ```suggestion
With the implementation above, the page content will be wrapped in a `ThemeProvider` component, so [MUI child components inherit the font family](https://github.com/vercel/next.js/discussions/45433) from the root layout.
```

Update the contents of `src/app/layout.tsx` to be:

{% raw %}
__app/layout.tsx__
```tsx
import WideLayout from "@/layout/WideLayout";
import { Ubuntu} from "next/font/google";

const font = Ubuntu({
  subsets: ['latin'],
  weight: ['300','400','500','700'],
});

export const metadata = {
  title: "Create Next App",
  description: "Generated by create next app",
};

export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  return (
    <html lang="en">
      <body className={font.className}>
        <WideLayout>{children}</WideLayout>
      </body>
    </html>
  );
}
```
{% endraw %}

> Also, remove `src/app/global.css` and `src/app/page.module.css`. Then run the client application with:

Reviewer: ```suggestion
Also, remove `src/app/globals.css` and `src/app/page.module.css`. Then run the client application with:
```

```shell
npm run dev

> ```

Reviewer: I get an error about Axios at this point.

<img width="985" alt="Screenshot 2023-09-08 at 8 47 30 AM" src="https://github.com/oktadev/okta-blog/assets/17892/70ce0524-5111-4eb6-98eb-e5a739e5aed7">

Author: I had a bad cache, added the missing import earlier in the post.

> Navigate to http://localhost:3000 and you should see the companies list:

Reviewer: ```suggestion
Navigate to `http://localhost:3000` and you should see the companies list:
```

 ▸    Successfully logged in.
 ▸    Tenant: dev-avup2laz.us.auth0.com
```

The next step is to create a client app, which you can do in one command:

```shell
auth0 apps create \
  --name "GraphQL Server" \
  --description "Spring Boot GraphQL Resource Server" \
  --type regular \
  --callbacks http://localhost:8080/login/oauth2/code/okta \
  --logout-urls http://localhost:8080 \
  --reveal-secrets
```

Once the app is created, you will see the OIDC app's configuration:

```
=== dev-avup2laz.us.auth0.com application created

  CLIENT ID            ***
  NAME                 GraphQL Server
  DESCRIPTION          Spring Boot GraphQL Resource Server
  TYPE                 Regular Web Application
  CLIENT SECRET        ***
  CALLBACKS            http://localhost:8080/login/oauth2/code/okta
  ALLOWED LOGOUT URLS  http://localhost:8080
  ALLOWED ORIGINS
  ALLOWED WEB ORIGINS
  TOKEN ENDPOINT AUTH
  GRANTS               implicit, authorization_code, refresh_token, client_credentials

 ▸    Quickstarts: https://auth0.com/docs/quickstart/webapp
 ▸    Hint: Emulate this app's login flow by running `auth0 test login ***`
 ▸    Hint: Consider running `auth0 quickstarts download ***`
```

> Add the `okta-spring-boot-starter` dependency to the `build.gradle` file:

Reviewer: ```suggestion
Add the `okta-spring-boot-starter` dependency to the `build.gradle` file in the spring-graphql-api project:
```

__build.gradle__
```groovy
dependencies {
    ...
    implementation 'com.okta.spring:okta-spring-boot-starter:3.0.5'
    ...
}
```

> Set the clientId, issuer, and audience for OAuth2 in the `application.properties` file:

Reviewer: ```suggestion
Set the client ID, issuer, and audience for OAuth 2.0 in the `application.properties` file:
```

__application.properties__
```properties
okta.oauth2.issuer=https://{yourAuth0Domain}/
okta.oauth2.client-id={clientId}

> okta.oauth2.audience=https://{yourAuth0Domain}/api/v2/

Reviewer: ```suggestion
okta.oauth2.audience=${okta.oauth2.issuer}api/v2/
```

```

Add the client secret to the `.env` file:

__.env__
```shell
export OKTA_OAUTH2_CLIENT_SECRET={clientSecret}
```

Add the following factory method to the class `SpringBootGraphQlApiConfig`, for requiring a bearer token for all requests:

__SpringBootGraphQlApiConfig.java__
```java
    ...
    @Bean
    public SecurityFilterChain configure(HttpSecurity http) throws Exception {
        http.oauth2ResourceServer(oauth2ResourceServer -> oauth2ResourceServer.jwt(withDefaults()));
        return http.build();
    }
    ...
```

> __NOTE:__

> > The Okta Spring Boot starter provides the security auto-configuration out of the box, and the resource server configuration would not be necessary. For some reason, the Spring for GraphQL cors allowed origins configuration does not take effect without the customization above.

Reviewer: ```suggestion
> The Okta Spring Boot starter provides the security auto-configuration out of the box, and the resource server configuration should not be necessary. For some reason, the Spring for GraphQL cors allowed origins configuration does not take effect without the customization above.
```

You are correct. I was able to verify this. Can you please enter an issue in the Okta Spring Boot starter project?

Author: I will entre the issue and link it here.

Again, in the root folder, run the API server with:

```shell
source .env && ./gradlew bootRun
```

> With HTTPie, send a request to the API server using a bearer access token:

Reviewer: You need to add a step to get the access token with the Auth0 CLI:

```
auth0 test token -a https://<your-auth0-domain>/api/v2/
```

Author: Added the the step

```shell
ACCESS_TOKEN={auth0AccessToken}
```

```shell
echo -E '{"query":"{\n    companyList(page: 20) {\n        id\n        SIC\n        name\n        status\n        category\n        companyNumber\n        countryOfOrigin\n    }\n}"}' | \
  http -A bearer -a $ACCESS_TOKEN POST http://localhost:8080/graphql
```

> __NOTE:__
> Follow [these instructions](https://auth0.com/docs/secure/tokens/access-tokens/get-management-api-access-tokens-for-testing) to get an access token for the Auth0 Management API, which can be used for testing the server API.

### Add Auth0 Login to the React client

> When using Auth0 as the identity provider, you can configure the Universal Login Page for a quick integration without having to build the login forms. First, create a SPA application using the Auth0 CLI:

Reviewer: ```suggestion
When using Auth0 as the identity provider, you can configure the Universal Login Page for a quick integration without having to build the login forms. First, register a SPA OIDC application using the Auth0 CLI:
```

```shell
auth0 apps create \
  --name "React client for GraphQL" \
  --description "SPA React client for a Spring GraphQL API" \
  --type spa \
  --callbacks http://localhost:3000/callback \
  --logout-urls http://localhost:3000 \

>   --reveal-secrets

Reviewer: Reveal secrets isn't needed here because a SPA shouldn't have a secret.

Reviewer: This prompts me for `allowed-origins`, so that should be passed in too. This command works w/o prompts:

```
auth0 apps create \
  --name "React client for GraphQL" \
  --description "SPA React client for a Spring GraphQL API" \
  --type spa \
  --callbacks http://localhost:3000/callback \
  --logout-urls http://localhost:3000 \
  --origins http://localhost:3000 \
  --web-origins http://localhost:3000
```

Author: Thanks, I updated the command.

```

> Copy the Auth0 domain and the client ID, and update the `src/.env.local` adding the following properties (add the new variables to the example file, too):

Reviewer: Why do they need to be added to the example file if it's not used?

Author: I added a comment about the example file earlier in the post.

__.env.local__
```shell
NEXT_PUBLIC_AUTH0_DOMAIN={yourAuth0Domain}
NEXT_PUBLIC_AUTH0_CLIENT_ID={clientId}
NEXT_PUBLIC_AUTH0_CALLBACK_URL=http://localhost:3000/callback
NEXT_PUBLIC_AUTH0_AUDIENCE=https://{yourAuth0Domain}/api/v2/
```

For handling the Auth0 post-login behavior, you need to add the page `src/app/callback/page.tsx` with the following content.

__callback/page.tsx__
```tsx
import Loader from "@/components/loader/Loader";

const Page = () => {
  return <Loader/>
};

export default Page;
```

For this example, the callback page will render empty.

Add the `@auth0/auth0-react` dependency to the project:

```shell

> npm install @auth0/auth0-react

Reviewer: Why aren't you using [nextjs-auth0](https://github.com/auth0/nextjs-auth0) here?

Author: Added the note.

  const onRedirectCallback = (appState?: AppState) => {
    router.push(appState?.returnTo || window.location.pathname);
  };

  if (!(domain && clientId && redirectUri)) {
    return null;
  }

  return (
    <Auth0Provider
      domain={domain}
      clientId={clientId}
      authorizationParams={{
        audience: audience,
        redirect_uri: redirectUri,
      }}
      useRefreshTokens={true}
      onRedirectCallback={onRedirectCallback}
    >
      <>{props.children}</>
    </Auth0Provider>
  );
};

export default Auth0ProviderWithNavigate;
```
{% endraw %}

The component `Auth0ProviderWithNavigate` wraps the children component with `Auth0Provider`, the provider of the Auth0 context, remembering the requested URL for redirection after login.
Use the component in the `WideLayout` component:

{% raw %}
__WideLayout.tsx__
```tsx
const WideLayout = (props: { children: React.ReactNode }) => {
  return (
    <ThemeProvider theme={theme}>
      <Auth0ProviderWithNavigate>
        <Container maxWidth="lg" sx={{ mt: 4 }}>
          {props.children}
        </Container>
      </Auth0ProviderWithNavigate>
    </ThemeProvider>
  );
};

> ```

Reviewer: I had to add the following import for this to compile:

```
import Auth0ProviderWithNavigate from '@/components/authentication/Auth0ProviderWithNavigate';
```

Author: I updated the block showing the whole file.

__AuthenticationGuard.tsx__
```tsx
'use client'

import { useAuth0 } from "@auth0/auth0-react";
import { useEffect } from "react";
import Loader from "../loader/Loader";

const AuthenticationGuard = (props: { children: React.ReactNode }) => {
  const { isLoading, isAuthenticated, error, loginWithRedirect } = useAuth0();

  useEffect(() => {
    if (!isAuthenticated && !isLoading) {
      loginWithRedirect({
        appState: { returnTo: window.location.href },
      });
    }
  }, [isAuthenticated, isLoading, loginWithRedirect]);

  if (isLoading) {
    return <Loader />;
  }
  if (error) {
    return <div>Oops... {error.message}</div>;
  }
  return <>{isAuthenticated && props.children}</>;
};

export default AuthenticationGuard;
```

The `AuthenticationGuard` component will be used to protect pages that require authentication, redirecting to the Auth0 universal login. Protect the index page by wrapping its content in the `AuthenticationGuard` component:

{% raw %}
__app/page.tsx__
```tsx
import AuthenticationGuard from "@/components/authentication/AuthenticationGuard";
import HomePage from "./HomePage";

const Page = () => {
  return (
    <AuthenticationGuard>

>       <HomePage></HomePage>

Reviewer: A recent version of React made it possible to do `<HomePage/>`

Author: Updated

        ...config,
        headers: {
          ...config.headers,
          Authorization: `Bearer ${accessToken}`,
        },
      };
    },
    function (error) {
      console.log("request interceptor error", error);
      return Promise.reject(error);
    }
  );
};
```

Add the file `src/hooks/useAccessToken.tsx` with the following content:

__useAccessToken.tsx__
```tsx
import { setInterceptors } from "@/services/auth";
import { useAuth0 } from "@auth0/auth0-react";
import { useCallback, useState } from "react";

export const useAccessToken = () => {
  const { isAuthenticated, getAccessTokenSilently } = useAuth0();
  const [accessToken, setAccessToken] = useState("");

  const saveAccessToken = useCallback(async () => {
    if (isAuthenticated) {
      try {
        const tokenValue = await getAccessTokenSilently();
        if (accessToken !== tokenValue) {
          setInterceptors(tokenValue);
          setAccessToken(tokenValue);
        }
      } catch (err) {
        // Inactivity timeout
        console.log("getAccessTokenSilently error", err);
      }
    }
  }, [getAccessTokenSilently, isAuthenticated, accessToken]);

  return {
    saveAccessToken,
  };
};
```

> The hook will call Auth0 `getAccessTokenSilently()` and will trigger a token refresh if the access token is expired. Then it will update axios interceptors to set the updated bearer token value in the request headers.

Reviewer: ```suggestion
The hook will call Auth0's `getAccessTokenSilently()` and trigger a token refresh if the access token expires. Then, it will update Axios interceptors to set the updated bearer token value in the request headers.
```

Create the `useAsyncWithToken` hook:

__useAsyncWithToken.tsx__
```tsx
import { useAccessToken } from "./useAccessToken";
import { useAsync } from "react-use-custom-hooks";

export const useAsyncWithToken = <T, P, E = string>(
  asyncOperation: () => Promise<T>, deps: any[]
) => {
  const { saveAccessToken } = useAccessToken();
  const [ data, loading, error ] = useAsync(async () => {
    await saveAccessToken();
    return asyncOperation();
  }, {},  deps);

  return {
    data,
    loading,
    error
  };
};
```

Update the calls in the `CompanyTableContainer` component to use the `useAsyncWithToken` hook:

__CompanyTableContainer.tsx__
```tsx
...
const {
  data: dataList,
  loading: loadingList,
  error: errorList,
} = useAsyncWithToken(
  () => CompanyApi.getCompanyList({ page: page - 1}),
  [props.page]
);

const { data: dataCount } = useAsyncWithToken(
  () => CompanyApi.getCompanyCount(),
  []
);

> ...

Reviewer: I had to add an import here:

```
import { useAsyncWithToken } from '@/hooks/useAsyncWithToken';
```

And this one is no longer needed:

```
import { useAsync } from 'react-use-custom-hooks';
```

Author: I am not sure how to explain the changes without showing the whole file again, I was trying to avoid repeating much the code. I re-formatted the block as diff.

```

Run the application with:

```shell
npm run dev
```

> Go to http://localhost:3000 and you should be redirected to the Auth0 universal login page. After login in, you should see the companies list again.

Reviewer: ```suggestion
Go to `http://localhost:3000` and you should be redirected to the Auth0 universal login page. After logging in, you should see the companies list again.
```

Reviewer: This took quite some time to load and redirect to the login page, around 10 seconds. Is that expected?

Reviewer: I also saw this error in my console:

```
FetchError: request to https://fonts.gstatic.com/s/ubuntu/v20/4iCs6KVjbNBYlgoKcQ72nU6AF7xm.woff2 failed, reason: getaddrinfo ENOTFOUND fonts.gstatic.com
    at ClientRequest.<anonymous> (/Users/mraible/Downloads/react-graphql/node_modules/next/dist/compiled/node-fetch/index.js:1:65756)
    at ClientRequest.emit (node:events:513:28)
    at TLSSocket.socketErrorListener (node:_http_client:502:9)
    at TLSSocket.emit (node:events:513:28)
    at emitErrorNT (node:internal/streams/destroy:151:8)
    at emitErrorCloseNT (node:internal/streams/destroy:116:3)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
  type: 'system',
  errno: 'ENOTFOUND',
  code: 'ENOTFOUND'
}
- error Failed to download `Ubuntu` from Google Fonts. Using fallback font instead.
```

Reviewer: After redirecting to `/callback`, I waited a minute before I gave up and refreshed the page. I opened a new incognito window and tried again. This time it took 20 seconds to load and redirect to the login page. Why is it so slow every time?

Reviewer: When it came back to `/callback`, I waited for 5 minutes before I gave up. It just spins.

<img width="1176" alt="Screenshot 2023-09-08 at 11 52 25 AM" src="https://github.com/oktadev/okta-blog/assets/17892/d138b46c-686e-48a9-b0bf-81918bf46b5c">

Author: > I also saw this error in my console:
> 
> ```
> FetchError: request to https://fonts.gstatic.com/s/ubuntu/v20/4iCs6KVjbNBYlgoKcQ72nU6AF7xm.woff2 failed, reason: getaddrinfo ENOTFOUND fonts.gstatic.com
>     at ClientRequest.<anonymous> (/Users/mraible/Downloads/react-graphql/node_modules/next/dist/compiled/node-fetch/index.js:1:65756)
>     at ClientRequest.emit (node:events:513:28)
>     at TLSSocket.socketErrorListener (node:_http_client:502:9)
>     at TLSSocket.emit (node:events:513:28)
>     at emitErrorNT (node:internal/streams/destroy:151:8)
>     at emitErrorCloseNT (node:internal/streams/destroy:116:3)
>     at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
>   type: 'system',
>   errno: 'ENOTFOUND',
>   code: 'ENOTFOUND'
> }
> - error Failed to download `Ubuntu` from Google Fonts. Using fallback font instead.
> ```

I also had issues and thought it was my networking. It seems there is some issue with [next/font](https://github.com/vercel/next.js/issues/45080). Will try a workaround.

{% img blog/spring-graphql-react/auth0-universal-login.png alt:"Auth0 universal login form" width:"400" %}{: .center-image }

{% img blog/spring-graphql-react/auth0-authorize-app.png alt:"Auth0 authorize application form" width:"400" %}{: .center-image }

Once the companies load, you can inspect the network requests and see the bearer token is sent in the request headers. It will look like the example below:

```

> Authorization: Bearer eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6IlJzZHNsM211SjNYU2ZVT0tDOEMxSiJ9.eyJodHRwczovL3d3dy5qaGlwc3Rlci50ZWNoL3JvbGVzIjpbIlJPTEVfQURNSU4iLCJST0xFX1VTRVIiXSwiaXNzIjoiaHR0cHM6Ly9kZXYtYXZ1cDJsYXoudXMuYXV0aDAuY29tLyIsInN1YiI6ImF1dGgwfDY0MzQxOTkxNTJmYjc2N2Y3ZWFlZDU2NyIsImF1ZCI6WyJhcGktc2VydmVyIiwiaHR0cHM6Ly9kZXYtYXZ1cDJsYXoudXMuYXV0aDAuY29tL3VzZXJpbmZvIl0sImlhdCI6MTY5MTU5NDE3NCwiZXhwIjoxNjkxNTk0Mjk0LCJhenAiOiI1WW5QeEJiTjRoYXFyd2JNaXRpZFBRVTBjb0l2YUI3SCIsInNjb3BlIjoib3BlbmlkIHByb2ZpbGUgZW1haWwgb2ZmbGluZV9hY2Nlc3MifQ.oRMJNIwSeO9dfHi6Q2tr_B51YetEPZqjcVEEIBe3ky9tEe50tTB5ssMTbVaR78_1qXA663Cn4EMPEYTLlb_wiOHEqnKpZnpq0O07G2MGszkrgv5giaQBOXvr9UT_Kc3pFPK-xMVmOsicLoF_mz8iyOzReG1Gcw0UbS1fZsJthdtC9svLiPGC1rn-dwPJxPpKy118vWLbEgO3NsdVfaiPxbucv0TL_B8Msd-wBD8N4M_yej9jl7w2JT0ltLza_-Glbxr1aQBKd19O2QT8ovgyE325BYqeYiUOaV7efwMgqHUm7z4LngaPLRNhMLP2BExG7bUQD2JjH2Mh19LFYNM-Gw

Reviewer: I like how this validates at jwt.io! :)

Author: Ooops, it shows the jhipster roles! I think I should make it invalid.

Author: I decided to just display the first characters.

    [props.page]
  );

  const { data: dataCount } = useAsyncWithToken(
    () => CompanyApi.getCompanyCount(),
    []
  );

  const onPageChange = (pagination: GridPaginationModel) => {
    const params = new URLSearchParams(searchParams.toString());
    const page = pagination.page + 1;
    params.set("page", page.toString());
    router.push(pathName + "?" + params.toString());
  };

  const companyData = dataList?.map((company: CompanyDTO) => {
    return {
      id: company.id,
      name: company.name,
      category: company.category,
      companyNumber: company.companyNumber,
      SIC: company.SIC,
      status: company.status,
      owner: company.controlledBy.map((person) => person.name).join(", "),
    }
  });

  return (
    <>
      {loadingList && <Loader />}
      {errorList && <div>Error</div>}

      {!loadingList && dataList && (
        <CompanyTable
          pagination={{ page: page - 1, pageSize: 10 }}
          rowCount={dataCount}
          rows={companyData}
          columns={columns}
          onPageChange={onPageChange}
        ></CompanyTable>
      )}
    </>
  );
};

export default CompanyTableContainer;
```
{% endraw %}

> Give it a try, and notice how simple was the update.

Reviewer: I don't think we want to say it was "simple" since it did require a fair bit of code. How about something like this?

> Give it a try. It's pretty neat how GraphQL allows you to get more data just by changing the client!

Author: I like it, I used that instead.

## Learn More About Spring Boot and React

> I hope you enjoyed this tutorial, and found this example useful. As you can see, not much work would be required to consume more company data from the GraphQL server, just a query update in the client. Also, the Auth0 Universal Login and Auth0 React SDK provide an efficient way to secure your React applications, following security best practices. You can find all the code for this example in the [Github repository](https://github.com/indiepopart/spring-graphql-react).

Reviewer: ```suggestion
I hope you enjoyed this tutorial, and found this example useful. As you can see, not much work would be required to consume more company data from the GraphQL server, just a query update in the client. Also, the Auth0 Universal Login and Auth0 React SDK provide an efficient way to secure your React applications, following security best practices. You can find all the code for this example in the [GitHub repository](https://github.com/oktadev/auth0-spring-graphql-react-example).
```

---
layout: blog_post

> title: "How to build a GraphQL API with Spring for GraphQL"

Reviewer: This ranks well (69) on our title analyzer. However, Spring Boot has more searches:

- spring graphql: 2400/month
- spring boot graphql: 3600/month

For this reason, I suggest we change it to "How to Build a GraphQL API with Spring Boot." This gets a 68, which is still pretty good.

> title: "How to build a GraphQL API with Spring for GraphQL"

Reviewer: ```suggestion
title: "How to Build a GraphQL API with Spring Boot"
```

- [PSCAmericans.csv](https://guides.neo4j.com/ukcompanies/data/PSCAmericans.csv)

Create the directory `src/main/docker` and create a file `neo4j.yml` there, with the following content:

__neo4j.yml__
```yml
name: companies
services:
  neo4j:
    image: neo4j:5
    volumes:
      - <csv-dir>:/var/lib/neo4j/import
    environment:
      - NEO4J_AUTH=neo4j/${NEO4J_PASSWORD}
      - NEO4JLABS_PLUGINS=["apoc"]
    # If you want to expose these ports outside your dev PC,
    # remove the "127.0.0.1:" prefix
    ports:
      - '127.0.0.1:7474:7474'
      - '127.0.0.1:7687:7687'
    healthcheck:
      test: ['CMD', 'wget', 'http://localhost:7474/', '-O', '-']
      interval: 5s
      timeout: 5s
      retries: 10
```

Create the file `src/main/docker/.env` with the following content:

__.env__
```dotenv
NEO4J_PASSWORD=verysecret
```

As you can see the compose file will mount `<csv-dir>` to a `/var/lib/neo4j/import` volume, making the content accessible from the running neo4j container.
Replace `<csv-dir>` with the path to the CSV files downloaded before.

In a terminal, go to the `docker` directory and run:

```shell
docker compose -f neo4j.yml up
```

> ### Run the API server

Reviewer: ```suggestion
### Run the Spring Boot API server
```

    []
  );

  const onPageChange = (pagination: GridPaginationModel) => {
    const params = new URLSearchParams(searchParams.toString());
    const page = pagination.page + 1;
    params.set("page", page.toString());
    router.push(pathName + "?" + params.toString());
  };

  const companyData = dataList?.map((company: CompanyDTO) => {
    return {
      id: company.id,
      name: company.name,
      category: company.category,
      companyNumber: company.companyNumber,
      SIC: company.SIC,
      status: company.status,
      owner: company.controlledBy.map((person) => person.name).join(", "),
    }
  });

  return (
    <>
      {loadingList && <Loader />}
      {errorList && <div>Error</div>}

      {!loadingList && dataList && (
        <CompanyTable
          pagination={{ page: page - 1, pageSize: 10 }}
          rowCount={dataCount}
          rows={companyData}
          columns={columns}
          onPageChange={onPageChange}
        ></CompanyTable>
      )}
    </>
  );
};

export default CompanyTableContainer;
```
{% endraw %}

Give it a try. It's pretty neat how GraphQL allows you to get more data just by changing the client!

> ## Learn More About Spring Boot and React

Reviewer: ```suggestion
## Learn More About Spring Boot, GraphQL, and React
```

I hope you enjoyed this tutorial, and found this example useful. As you can see, not much work would be required to consume more company data from the GraphQL server, just a query update in the client. Also, the Auth0 Universal Login and Auth0 React SDK provide an efficient way to secure your React applications, following security best practices. You can find all the code for this example in the [GitHub repository](https://github.com/oktadev/auth0-spring-graphql-react-example).

> Check out the Auth0 documentation for adding [sign-up](https://developer.auth0.com/resources/guides/spa/react/basic-authentication#add-user-sign-up-to-react) and [logout](https://developer.auth0.com/resources/guides/spa/react/basic-authentication#add-user-logout-to-reactfunctionality) to your React application. And for more fun tutorials about Spring Boot and React, you can visit the following links:

Reviewer: ```suggestion
Check out the Auth0 documentation for adding [sign-up](https://developer.auth0.com/resources/guides/spa/react/basic-authentication#add-user-sign-up-to-react) and [logout](https://developer.auth0.com/resources/guides/spa/react/basic-authentication#add-user-logout-to-reactfunctionality) to your React application. And for more fun tutorials about Spring Boot, GraphQL, and React, you can visit the following links:
```

- [Build a Simple CRUD App with Spring Boot and Vue.js](https://auth0.com/blog/build-crud-spring-and-vue/)
- [Use React and Spring Boot to Build a Simple CRUD App](https://auth0.com/blog/simple-crud-react-and-spring-boot/)

> - [Full Stack Java with React, Spring Boot, and JHipster](https://auth0.com/blog/full-stack-java-with-react-spring-boot-and-jhipster/)

Reviewer: ```suggestion
- [The Complete Guide to React User Authentication with Auth0](https://auth0.com/blog/complete-guide-to-react-user-authentication/)
- [Build and Secure a GraphQL Server with Node.js](https://auth0.com/blog/build-and-secure-a-graphql-server-with-node-js/)
- [Full Stack Java with React, Spring Boot, and JHipster](https://auth0.com/blog/full-stack-java-with-react-spring-boot-and-jhipster/)
```

✔ Would you like to customize the default import alias? ... No
```

> Then add the the MUI Datagrid dependencies, custom hooks from Vercel and Axios:

Reviewer: ```suggestion
Then add the the MUI Datagrid dependencies, custom hooks from Vercel, and Axios:
```
Oxford comma.

source .env && ./gradlew bootRun
```

> Get an access token using the Auth0 cli with the `test token` command:

Reviewer: ```suggestion
Get an access token using the Auth0 CLI with the `auth0 test token` command:
```

```shell

> auth0 test token -a https://<yourAuth0Domain>/api/v2/

Reviewer: Is `<yourAuth0Domain>` the same placeholder you've used in other places? It doesn't matter what we use, as long as we're consistent.

Author: Yes, it is the same. Is there a convention? I thought I was using the right convention as for Okta.

Author: I updated the placeholders to use the hyphen
